stage #4/module #5. AWS/theory/dynamodb.md

```diff
@@ -0,0 +1,143 @@
+DynamoDB uses the partition key value as input to an internal hash function. The output from the hash function determines the partition (physical storage internal to DynamoDB) in which the item will be stored. All items with the same partition key value are stored together, in sorted order by sort key value.<br/>
+In a table that has a partition key and a sort key, it's possible for multiple items to have the same partition key value. However, those items must have different sort key values.<br/>
+The Music table described in Tables, items, and attributes is an example of a table with a composite primary key (Artist and SongTitle). You can access any item in the Music table directly, if you provide the Artist and SongTitle values for that item.<br/>
+A composite primary key gives you additional flexibility when querying data. For example, if you provide only the value for Artist, DynamoDB retrieves all of the songs by that artist. To retrieve only a subset of songs by a particular artist, you can provide a value for Artist along with a range of values for SongTitle.<br/>
+*The partition key of an item is also known as its hash attribute. The term hash attribute derives from the use of an internal hash function in DynamoDB that evenly distributes data items across partitions, based on their partition key values.
+The sort key of an item is also known as its range attribute. The term range attribute derives from the way DynamoDB stores items with the same partition key physically close together, in sorted order by the sort key value.*
+
+Secondary indexes.<br/>
+You can create one or more secondary indexes on a table. A secondary index lets you query the data in the table using an alternate key, in addition to queries against the primary key. DynamoDB doesn't require that you use indexes, but they give your applications more flexibility when querying your data. After you create a secondary index on a table, you can read data from the index in much the same way as you do from the table.
+<br/>The service supports two kinds of indexes<br/>
+1. Global secondary index – An index with a partition key and sort key that can be different from those on the table.
+2. Local secondary index – An index that has the same partition key as the table, but a different sort key.<br/>
+
+*You can think of these the following way: local secondary index defines distinctness within a single partition key whereas global one is about distinctness within a single pair of partition key and sort one*
+Each table in DynamoDB has a quota of 20 global secondary indexes (default quota) and 5 local secondary indexes.
+
+DynamoDB Streams is an optional feature that captures data modification events in DynamoDB tables. The data about these events appear in the stream in near-real time, and in the order that the events occurred.
```

Comment:
This paragraph is okay, but its formatting is a bit dangerous to ordinary people. There are no <br /> breaks and there is quite a lot of text. Could you add some, pls?

Reply:
fixed